File: src/safe.ts
export const SAFE_FILTER = Symbol.for("selmer.safe-filter");
export const SAFE_CONTEXT_KEY = "__selmerSafeFilter";

export interface SafeValue {
  readonly [SAFE_FILTER]: true;
  readonly value: unknown;
}

export function safe(value: unknown): SafeValue {
  return { [SAFE_FILTER]: true, value };
}

export function isSafeValue(value: unknown): value is SafeValue {
  return Boolean(value && typeof value === "object" && (value as SafeValue)[SAFE_FILTER]);
}

export function unwrapSafe(value: unknown): unknown {
  if (isSafeValue(value)) return value.value;
  if (Array.isArray(value) && (value[0] === ":safe" || value[0] === "safe")) return value[1];
  return value;
}

File: src/selmer/safe.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

SAFE_CONTEXT_KEY = "__selmer_safe_filter__"


@dataclass(frozen=True)
class SafeValue:
    value: Any


def safe(value: Any) -> SafeValue:
    return SafeValue(value)


def is_safe_value(value: Any) -> bool:
    return isinstance(value, SafeValue) or (
        isinstance(value, (list, tuple)) and len(value) >= 2 and value[0] in {":safe", "safe"}
    )


def unwrap_safe(value: Any) -> Any:
    if isinstance(value, SafeValue):
        return value.value
    if isinstance(value, (list, tuple)) and len(value) >= 2 and value[0] in {":safe", "safe"}:
        return value[1]
    return value
